src/geometry.rs

```diff
@@ -447,86 +361,186 @@ impl Hash for UnorientedEdge {
     }
 }
 
-#[allow(dead_code, clippy::type_complexity)]
 fn remove_orphan_vertices(
-    faces: Vec<(u32, u32, u32)>,
-    vertices: Vec<Point3<f32>>,
-) -> (Vec<(u32, u32, u32)>, Vec<Point3<f32>>) {
-    let mut old_to_new_vertex_index: HashMap<u32, u32> = HashMap::new();
-    let mut reduced_vertices: Vec<Point3<f32>> = Vec::new();
-    let mut new_index = 0;
-    #[allow(clippy::needless_range_loop)]
-    for original_index in 0..vertices.len() {
-        let o_i = cast_u32(original_index);
-        if faces.iter().any(|f| f.0 == o_i || f.1 == o_i || f.2 == o_i) {
-            old_to_new_vertex_index.insert(o_i, new_index);
-            reduced_vertices.push(vertices[original_index]);
-            new_index += 1;
-        }
-    }
-
-    let recomputed_faces: Vec<_> = faces
-        .iter()
-        .map(|f| {
-            (
-                *old_to_new_vertex_index
-                    .get(&f.0)
-                    .expect("Vertex index not found"),
-                *old_to_new_vertex_index
-                    .get(&f.1)
-                    .expect("Vertex index not found"),
-                *old_to_new_vertex_index
-                    .get(&f.2)
-                    .expect("Vertex index not found"),
-            )
-        })
-        .collect();
-
-    (recomputed_faces, reduced_vertices)
+    faces: FaceVertexTuples,
+    vertices: Vertices,
+) -> (FaceVertexTuples, Vertices) {
+    let mut vertices_reduced: Vertices = Vec::new();
+    let original_vertex_len = vertices.len();
+    let unused_vertex_marker = vertices.len();
+    let mut old_new_vertex_map: Vec<usize> = vec![unused_vertex_marker; original_vertex_len];
+    let mut faces_renumbered: FaceVertexTuples = Vec::new();
+    for face in faces {
+        let old_vertex_index_0 = cast_usize(face.0);
+        let new_vertex_index_0 = if old_new_vertex_map[old_vertex_index_0] == unused_vertex_marker {
+            vertices_reduced.push(vertices[old_vertex_index_0]);
+            let new_index = vertices_reduced.len() - 1;
```

Comment:
I would move this `let new_index =` one line above, and get rid of the `- 1`. The current form triggered the underflow detector in my brain 👻 . Same for all analogous code in this functions and the others.